File: api/market/src/services-impl/loggers/RemoteLogger.ts
import fluentLogger from 'fluent-logger';
import {ILogger} from "../../Services";
import {ValidationError} from "../../consts/ServerError";
import {AppStage} from "./AppStage";

let internalRemoteLogger : InternalRemoteLogger; // caching

export interface RemoteLoggerInitData {
    serviceName: string;
    instanceId?: string;
    stage: AppStage;
    remoteHost?: string;
}

export class RemoteLogger implements ILogger {//
    private readonly _logger?: ILogger;
    private readonly _fallback?: ILogger;

    constructor(
        private readonly _initData: RemoteLoggerInitData,
        fallback?: ILogger,
        private readonly _zoneName?: string,
    ) {
        if (_initData.stage === 'local') {
            this._fallback = fallback;
        }
        if (!internalRemoteLogger) {
            if (_initData.remoteHost) {
                const [host, portStr] = _initData.remoteHost.split(':');
                const port = portStr ? parseInt(portStr, 10) : 0;
                if (port > 0) {
                    internalRemoteLogger = this._logger = new InternalRemoteLogger(_initData, _zoneName, host, port);
                }
            }
        } else {
            this._logger = internalRemoteLogger;
        }
    }

    info(message: any): void {
        this._fallback?.info(message);
        this._logger?.info(message);
    }

    infos(...message: any[]): void {
        this._fallback?.infos(message);
        this._logger?.infos(message);
    }

    error(err: any): void {
        this._fallback?.error(err);
        this._logger?.error(err);
    }

    errors(...err: any[]): void {
        this._fallback?.errors(err);
        this._logger?.errors(err);
    }

    assert(condition: any, message: string): void {
        if (!condition) {
            this._fallback?.error(message);
            this._logger?.error(message);
            throw new ValidationError(message);
        }
    }

    clone(prefix: string): ILogger {
        if (this._fallback) {
            return new RemoteLogger(this._initData, this._fallback.clone(prefix), this._zoneName);
        }
        return new RemoteLogger(this._initData, undefined, this._zoneName);
    }
}

let fluent: fluentLogger.FluentSender<unknown>;

function createFluent(tag: string, host: string, port: number) {
    if (!fluent) {
        fluent = fluentLogger.createFluentSender(tag, {host, port});
        console.log(`Use FluentLogger with tag=${tag} host=${host} port=${port}`);
    }
    return fluent;
}

class InternalRemoteLogger implements ILogger {
    private logger: fluentLogger.FluentSender<unknown>;
    private constructorHelper: LogMessageConstructor;

    constructor(
        initData: RemoteLoggerInitData,
        zoneName: string | undefined,
        host: string,
        port: number,
    ) {
        this.constructorHelper = new LogMessageConstructor(initData, zoneName);
        const tag = `service.${this.constructorHelper.tag}`;
        this.logger = createFluent(tag, host, port);
    }

    info(message: any): void {
        const m = this.turnObjIntoString(message);
        this.logger.emit('event', this.constructorHelper.info(m));
    }

    infos(...message: any[]): void {
        const arr: string[] = [];
        message.forEach(e => arr.push(this.turnObjIntoString(e)));
        const m = arr.join(', ');
        this.logger.emit('event', this.constructorHelper.info(m));
    }

    error(message: string) {
        this.logger.emit('event', this.constructorHelper.error(message));
    }

    errors(...err: any[]): void {
        const arr: string[] = [];
        err.forEach(e => arr.push(this.turnObjIntoString(e)));
        const m = arr.join(', ');
        this.logger.emit('event', this.constructorHelper.error(m));
    }

    assert(condition: any, message: string): void {
        throw new Error('Method not implemented.');
    }

    clone(prefix: string): ILogger {
        return this;
    }

    private turnObjIntoString(obj: any): string {
        const t = typeof obj;
        if (t === 'string') {
            return obj;
        }
        if (t === 'number' || t === 'boolean') {
            return obj.toString();
        }
        if (t === 'object') {
            if (obj instanceof Error) {
                const ex = obj;
                const stack = ex.stack?.split('\n').slice(0, 5).join('\n') || '';
                return `${ex.message}\n${stack}`;
            }
            return JSON.stringify(obj);
        }
        throw new Error(`Unknown type of ${t}`);
    }
}

class LogMessageConstructor {
    public tag: string;
    private readonly _stageName: string;
    private readonly _instanceId?: string;

    constructor(
        initData: RemoteLoggerInitData,
        private readonly _zoneName?: string,
    ) {
        this._stageName = initData.stage.toLowerCase();
        this._instanceId = initData.instanceId;
        this.tag = `${initData.serviceName}.${this._stageName}${this._instanceId ? '.' + this._instanceId : ''}`;
    }

    private constructLog(message: string, level: string) {
        const data: Record<string, string> = {
            message,
            level,
            service: this.tag,
            stage: this._stageName,
        };

        if (this._instanceId) data['instance_id'] = this._instanceId;
        if (this._zoneName) data['zone'] = this._zoneName;

        return data;
    }

    info(message: string) {
        return this.constructLog(message, 'info');
    }

    error(message: string) {
        return this.constructLog(message, 'error');
    }
}

File: server/Common/src/com/senspark/common/utils/RemoteLogger.kt
package com.senspark.common.utils

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.fluentd.logger.FluentLogger
import java.util.Base64

enum class AppStage {
    LOCAL, TEST, PROD;

    companion object {
        fun fromString(value: String): AppStage {
            return when (value.lowercase()) {
                "local" -> LOCAL
                "test" -> TEST
                "prod" -> PROD
                else -> throw IllegalArgumentException("Unknown stage: $value")
            }
        }

        fun toString(stage: AppStage): String {
            return when (stage) {
                LOCAL -> "local"
                TEST -> "test"
                PROD -> "prod"
            }
        }
    }
}

data class RemoteLoggerInitData(
    val serviceName: String,
    val instanceId: String?,
    val stage: AppStage,
    val remoteHost: String,
)

class RemoteLogger(
    initData: RemoteLoggerInitData,
    zoneName: String?,
    allowFallback: Boolean = true,
) : ILogger, IGlobalLogger, IServerLogger {

    private val _logger = mutableListOf<ILogger>()
    private val _scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    init {
        val h = initData.remoteHost.split(":")
        val host = h.first()
        if (initData.stage == AppStage.LOCAL) {
            _logger.add(PrintLogger())
        }
        if (host != "localhost") {
            val port = if (h.size > 1) {
                h[1].toInt()
            } else 0
            if (port > 0) {
                _logger.add(InternalRemoteLogger(initData, zoneName, host, port))
            }
        }
    }

    override fun log2(visibleTag: String, compressibleMessage: String, customColor: ColorCode) {
        _scope.launch {
            _logger.forEach { it.log2(visibleTag, compressibleMessage, customColor) }
        }
    }

    override fun log2(visibleTag: String, compressibleMessage: () -> String, customColor: ColorCode) {
        _scope.launch {
            _logger.forEach { it.log2(visibleTag, compressibleMessage, customColor) }
        }
    }

    override fun log(message: String, customColor: ColorCode) {
        _scope.launch {
            _logger.forEach { it.log(message, customColor) }
        }
    }

    override fun warn(message: String) {
        _scope.launch {
            _logger.forEach { it.warn(message) }
        }
    }

    override fun error(message: String) {
        _scope.launch {
            _logger.forEach { it.error(message) }
        }
    }

    override fun error(ex: Exception) {
        _scope.launch {
            _logger.forEach { it.error(ex) }
        }
    }

    override fun error(prefix: String, ex: Exception) {
        _scope.launch {
            _logger.forEach { it.error(prefix, ex) }
        }
    }

    override fun initialize() {
    }
}

private object StaticRemoteLogger {
    private lateinit var _logger: FluentLogger

    fun getLogger(tag: String, host: String, port: Int): FluentLogger {
        if (!::_logger.isInitialized) {
            _logger = FluentLogger.getLogger(tag, host, port)
            println("Use FluentLogger with tag=${tag} host=${host} port=${port}")
        }
        return _logger
    }
}

private class InternalRemoteLogger(
    initData: RemoteLoggerInitData,
    zoneName: String?,
    host: String,
    port: Int,
) : ILogger {
    private val _logMsgConstructor = LogMessageConstructor(initData, zoneName)
    private val _logger = StaticRemoteLogger.getLogger("service.${_logMsgConstructor.tag}", host, port)

    companion object {
        private const val MAX_MESSAGE_LENGTH = 1000
        private const val EVENT_TYPE = "event"
    }

    override fun log2(visibleTag: String, compressibleMessage: String, customColor: ColorCode) {
        val msg = if (compressibleMessage.length > MAX_MESSAGE_LENGTH) {
            compressibleMessage.toByteArray(Charsets.UTF_8).let { bytes ->
                val outputStream = java.io.ByteArrayOutputStream()
                java.util.zip.GZIPOutputStream(outputStream).use { gzip ->
                    gzip.write(bytes)
                }
                Base64.getEncoder().encodeToString(outputStream.toByteArray())
            }
        } else {
            compressibleMessage
        }
        _logger.log(EVENT_TYPE, _logMsgConstructor.info("$visibleTag $msg"))
    }

    override fun log2(visibleTag: String, compressibleMessage: () -> String, customColor: ColorCode) {
        val msg = compressibleMessage()
        log2(visibleTag, msg, customColor)
    }

    override fun log(message: String, customColor: ColorCode) {
        _logger.log(EVENT_TYPE, _logMsgConstructor.info(message))
    }

    override fun warn(message: String) {
        _logger.log(EVENT_TYPE, _logMsgConstructor.warn(message))
    }

    override fun error(message: String) {
        _logger.log(EVENT_TYPE, _logMsgConstructor.error(message))
    }

    override fun error(ex: Exception) {
        _logger.log(EVENT_TYPE, _logMsgConstructor.error(null, ex))
    }

    override fun error(prefix: String, ex: Exception) {
        _logger.log(EVENT_TYPE, _logMsgConstructor.error(prefix, ex))
    }
}

/**
 * Cấu trúc log tag: serviceName.stage.instanceId?
 * Ví dụ:
 * - sv-bomb.local.sv-game-v1
 * - sv-bomb.test.sv-game-v2
 * - sv-bomb.prod.sv-pvp-v1
 * - ap-login.test
 *
 * Service name: tag
 */
private class LogMessageConstructor(
    initData: RemoteLoggerInitData,
    private val zoneName: String?
) {
    val tag: String
    private val _stageName = initData.stage.name.lowercase()
    private val instanceId = initData.instanceId

    init {
        var tag = "${initData.serviceName}.${_stageName}"
        if (instanceId != null) {
            tag += ".$instanceId"
        }
        this.tag = tag
    }

    fun error(errorMessage: String): Map<String, String> {
        return constructLog(errorMessage, "error")
    }

    fun error(prefix: String?, exception: Exception): Map<String, String> {
        val stackTrace = exception.stackTraceToString()
        val maxLines = 5
        val stackTraceLines = stackTrace.split("\n")
        val limitedStackTrace = stackTraceLines.take(maxLines).joinToString("\n")
        if (prefix != null) {
            val errorMessage = "$prefix: ${exception.message}\n$limitedStackTrace"
            return error(errorMessage)
        }
        return error(limitedStackTrace)
    }

    fun info(message: String): Map<String, String> {
        return constructLog(message, "info")
    }

    fun warn(message: String): Map<String, String> {
        return constructLog(message, "warn")
    }

    private fun constructLog(message: String, level: String): Map<String, String> {
        val data = mutableMapOf(
            "message" to message,
            "level" to level,
            "service" to tag,
            "stage" to _stageName,
        )
        if (instanceId != null) {
            data["instance_id"] = instanceId
        }
        if (zoneName != null) {
            data["zone"] = zoneName
        }
        return data
    }
}